### PR TITLE
DrivAerML: T_max=24 narrow schedule sweep

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Shorter cosine cycle (T_max=24 vs champion T_max=30) means the LR returns to minimum ~20% faster, giving more "warm restart" cycles within the 600-minute budget. If the model benefits from more frequent restarts rather than longer warmups, this should improve convergence.

## Bucket
E — Cosine Schedule Sweep

## Command
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=600 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 24 --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 --no-compile-model --wandb_group dm-narrow-schedule
```

## Kill Condition
If clearly worse than champion-control (3.833%) by epoch 160 with no improvement trend, kill.

## Baseline
- **val surface_rel_l2_pct:** 3.833% at epoch 511
- **test surface_rel_l2_pct:** 4.685%
- **W&B run:** ncl1dh88 (eren/dm-ema-9995-gc05)
- **Full-eval test evidence:** ~4.39-4.50% (see issue #3283)
- **Target:** <3.71% (AB-UPT benchmark)

## Expected outcome
T_max=24 and T_max=36 (chrome) together bracket the champion T_max=30. Best result informs a refined schedule sweep.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nami:

## Results

### DrivAerML T_max=24 Narrow Schedule Sweep

| Metric | T_max=24 (this run) | Champion T_max=30 | Delta | Target |
|--------|---------------------|-------------------|-------|--------|
| **Val surface_rel_l2_pct** | 4.093% (epoch 825) | 3.833% (epoch 511) | +0.260pp ❌ | <3.71% |
| **Test surface_rel_l2_pct** | 4.578% | 4.685% | **-0.107pp** ✅ | — |

### Command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=600 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 24 --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 --no-compile-model --wandb-group dm-narrow-schedule --agent nami --wandb-name "nami/dm-tmax24-sweep"
```

### W&B
- **Run ID:** qwk5wg13
- **Run name:** nami/dm-tmax24-sweep
- **Group:** dm-narrow-schedule
- **Total epochs:** 881 (600-minute timeout)

### What happened

**Val metric — hypothesis falsified.** T_max=24 converged to 4.093% val, 0.26pp worse than the champion's 3.833%. The shorter cosine cycle did not improve validation convergence.

**Test metric — surprising generalization improvement.** Despite the worse val metric, T_max=24 achieved 4.578% test, which is 0.107pp *better* than the champion's 4.685%. This suggests T_max=30 may be slightly overfitting to the validation set, while T_max=24's more frequent warm restarts produce better generalization.

**Training dynamics were distinctive:**
- **Epochs 1–414:** Steady descent to 4.226% val, then entered a ~100-epoch plateau
- **Epochs 520–570:** Second descent wave, breaking to 4.205%
- **Epochs 645–670:** Third descent wave to 4.181%
- **Epochs 759–767:** Fourth descent wave to 4.130%
- **Epochs 823–825:** Fifth descent wave to final best 4.093%
- **Epoch 860:** Brief disruption (loss spike to 0.011) followed by recovery

The model exhibited a distinctive "staircase" convergence pattern — long plateaus punctuated by sudden drops. Each cosine cycle trough (lr=0) was where improvements typically appeared. The model made significant gains well past epoch 600, unlike the champion which found its best at epoch 511.

**More cosine cycles = more opportunities for late breakthrough.** With T_max=24, the model completed ~37 full cosine cycles vs ~29 for T_max=30 in the same wall-clock budget. The extra 8 cycles gave more chances for the lr minimum to trigger descent, which is likely why the model kept finding improvements so late.

### Suggested follow-ups

1. **Investigate the val/test divergence.** The fact that T_max=24 has worse val but better test than the champion is noteworthy. This could suggest the champion's schedule overfits slightly to the val split, or that more frequent restarts act as regularization.

2. **Try T_max=20 or T_max=18.** If shorter cycles improve test generalization, even shorter cycles might push this further. However, there's likely a floor where cycles become too short to converge within each cycle.

3. **Ensemble T_max=24 and T_max=30 checkpoints.** Since they appear to learn slightly different features (different val/test trade-off), ensembling might yield a better overall result.

4. **Run with compile_model=True.** This run used --no-compile-model. With compilation, each epoch would be faster, potentially fitting more training into the 600-minute budget.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3297 because ADVISOR: Closing — T_max=24 val=4.093% doesn't beat any baseline (3.622% T_max=36, 3.700% metric-aware, 3.833% T_max=30). T_max=24 is too short. The schedule sweep confirms T_max=36 > T_max=30 > T_max=24. Nami reassigned..
